server/core/src/https/middleware/mod.rs

```diff
@@ -0,0 +1,83 @@
+use axum::{
+    http::{self, Request},
+    middleware::Next,
+    response::Response,
+    Extension,
+};
+use http::{HeaderMap, HeaderValue};
+use uuid::Uuid;
+
+pub mod compression;
+
+// the version middleware injects
+const KANIDM_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+/// Injects a header into the response with "X-KANIDM-VERSION" matching the version of the package.
+pub async fn version_middleware<B>(request: Request<B>, next: Next<B>) -> Response {
+    let mut response = next.run(request).await;
+    let headers = response.headers_mut();
+    headers.insert(
+        "X-KANIDM-VERSION",
+        HeaderValue::from_str(KANIDM_VERSION).unwrap(),
+    );
+
+    response
+}
+
+#[derive(Clone)]
+/// For holding onto the event ID
+pub struct KOpId {
+    pub eventid: Uuid,
+    pub value: String,
+    pub uat: Option<String>,
+}
+
+/// This runs at the start of the request, adding an extension with the OperationID
+pub async fn kopid_start<B>(
+    headers: HeaderMap,
```

Comment:
Maybe use a TypedHeader for auth token extraction?

https://docs.rs/axum/0.6.18/axum/struct.TypedHeader.html#as-extractor

with this

https://docs.rs/headers/0.3.8/headers/authorization/struct.Bearer.html

Would that work or have I missed something?

Reply:
yep, it's on my list of things to go back and do :smile: added it as an explicit todo now. I can't just use an extractor because that *enforces* the header, and I just tried to do it as an optional but then axum got grumpy

Reply:
Hi @yaleman,
looks like I messed up the "demo" PR for you.
I did an optional auth header extraction here https://github.com/kanidm/kanidm/commit/c090310f2df1f9dd4afc72b6aee0eb908adaced8

The build works as some tests are still failing it is hard for me to reason why. Hopefully, this is still useful to you.

Reply:
Thanks for that, I'll give it a go - I wouldn't say you've messed anything up at all! 😄

Reply:
Woo, that seems to work, thanks!